lti_consumer/lti_consumer.py

```diff
@@ -0,0 +1,839 @@
+        (Note: this prevents good debug messages for the client, so we might want to change this, or the spec)
+
+        Arguments:
+            request (xblock.django.request.DjangoWebobRequest):  Request object for current HTTP request
+            suffix (unicode):  request path after "lti_2_0_result_rest_handler/".  expected to be "user/<anon_id>"
+
+        Returns:
+            webob.response:  response to this request.  See above for details.
+        """
+        lti_consumer = Lti2Consumer(self)
+
+        if self.runtime.debug:
+            lti_provider_key, lti_provider_secret = self.lti_provider_key_secret
+            log_authorization_header(request, lti_provider_key, lti_provider_secret)
+
+        if not self.accept_grades_past_due and self.is_past_due:
+            return Response(status=404)  # have to do 404 due to spec, but 400 is better, with error msg in body
+
+        try:
+            anon_id = parse_handler_suffix(suffix)
+        except LtiError:
+            return Response(status=404)  # 404 because a part of the URL (denoting the anon user id) is invalid
+        try:
+            lti_consumer.verify_result_headers(request, verify_content_type=True)
+        except LtiError:
+            return Response(status=401)  # Unauthorized in this case.  401 is right
+
+        user = self.runtime.get_real_user(anon_id)
+        if not user:  # that means we can't save to database, as we do not have real user id.
+            msg = "[LTI]: Real user not found against anon_id: {}".format(anon_id)
+            log.info(msg)
+            return Response(status=404)  # have to do 404 due to spec, but 400 is better, with error msg in body
+
+        try:
+            # Call the appropriate Lti2Consumer method
+            args = []
+            if request.method == 'PUT':
+                # Request body should be passed as an argument
+                # to result handler method on PUT
+                args.append(request.body)
+            response_body = getattr(lti_consumer, "{}_result".format(request.method.lower()))(user, *args)
+        except (AttributeError, LtiError):
+            return Response(status=404)
+
+        return Response(
+            json.dumps(response_body),
+            content_type=Lti2Consumer.CONTENT_TYPE_RESULT_JSON,
+        )
+
+    def max_score(self):
+        """
+        Returns the configured number of possible points for this component.
+
+        Arguments:
+            None
+
+        Returns:
+            float: The number of possible points for this component
+        """
+        return self.weight if self.has_score else None
+
+    def clear_user_module_score(self, user):
+        """
+        Clears the module user state, including grades and comments, and also scoring in db's courseware_studentmodule
+
+        Arguments:
+            user (django.contrib.auth.models.User):  Actual user whose module state is to be cleared
+
+        Returns:
+            nothing
+        """
```

Comment:
Thanks for all these docstrings! You might consider using PyContracts in the future - just search for `@contract` in the current edx-platform codebase for examples.